### PR TITLE
fix(pagination): revamp pagination permalink management

### PR DESCRIPTION
**Pull request by Geobert**
fixes https://github.com/cobalt-org/cobalt.rs/issues/622

### CODE REVIEW

File: src/cobalt_model/permalink.rs [hunk 1]
```diff
@@ -24,7 +24,7 @@ fn explode_permalink_string(permalink: &str, attributes: &liquid::value::Object)
     // Handle cases where substutions were blank
     p = p.replace("//", "/");
 
-    if p.starts_with('/') {
+    if p.starts_with('/') && p.len() > 1 {
```
Comment (Geobert) [inline src/cobalt_model/permalink.rs]: I've added this to avoid empty `doc.url_path` which broke some pagination navigation

File: src/pagination/mod.rs [hunk 2]
```diff
@@ -136,23 +135,45 @@ fn interpret_permalink(
...
+    let permalink_path = std::path::Path::new(&permalink);
+    let mut pagination_root = permalink_path.extension().map_or_else(
+        || permalink.clone(),
+        |os_str| {
+            permalink
+                .trim_end_matches(&format!(".{}", os_str.to_string_lossy()))
+                .to_string()
+        },
+    );
+    // avoid "//"
```
Comment (Geobert) [inline src/pagination/mod.rs]: `//all/_p/2/` is not working, it gives us a link `all/_p/2` which leads to `http://all/_p/2/`.

Whereas `/all/_p/2/` gives us `http://localhost:3030/all/_p/2/`

**Comment by Geobert** [comment]
Hm, I should have test on my blog before opening the PR, I have a big issue. I'll update the tests to reflect the bug

**Comment by Geobert** [comment]
I don't understand something while serving with cobalt:
`<a href="tags/tag1/_p/2"` gives me a wrong url `http://localhost:3000/tags/tag1/tags/tag1/_p/2` (note the double `/tags/tag1/`)

whereas `<a href="/tags/tag1/_p/2"` (with a leading slash) gives me `http://localhost:3000/tags/tag1/_p/2` which is correct. 

Why is that?

File: src/pagination/mod.rs (hunk 2 above)
Comment (epage) [reply src/pagination/mod.rs]: We do not support user-configured relative paths out of simplicity *but* the post-processed permalinks are made relative in case the generated site is not in the user's root.

I recommend doing an explicit `/{{ page.permalink}}` or even better `{{ site.base_url }}/{{ page.permalink }}`.

**Comment by epage** [comment]
> `<a href="tags/tag1/_p/2"`

This is a relative path, so where it leads depends on what page in your site you are starting from.  I assume this link is being generated in `tags/tag1/index.html` and following that leads to `tags/tag1/tags/tag1/_p/2`?

File: src/pagination/mod.rs (hunk 2 above)
Comment (Geobert) [reply src/pagination/mod.rs]: Ok I see! Make sense. Removing that!

File: src/cobalt_model/pagination_config.rs [hunk 3]
```diff
@@ -121,13 +121,15 @@ impl PaginationConfigBuilder {
...
-        let permalink = permalink.unwrap_or_else(|| DEFAULT_PERMALINK.to_owned());
+        let permalink_suffix =
+            permalink_suffix.unwrap_or_else(|| DEFAULT_PERMALINK_SUFFIX.to_owned());
         let order = order.unwrap_or(SortOrder::Desc);
         let sort_by = sort_by.unwrap_or_else(|| vec![DEFAULT_SORT.to_owned()]);
         Some(PaginationConfig {
             include,
             per_page,
-            permalink,
+            front_permalink: permalink.to_string(),
```
Comment (epage) [inline src/cobalt_model/pagination_config.rs]: `to_string` I assume your intention is more so `to_owned`?

They both work, the latter just clarifies intent

File: src/cobalt_model/permalink.rs (hunk 1 above)
Comment (epage) [reply src/cobalt_model/permalink.rs]: Any idea why?  If so, we should have test cases to clarify..

**Comment by Geobert** [comment]
> 
> 
> > `<a href="tags/tag1/_p/2"`
> 
> This is a relative path, so where it leads depends on what page in your site you are starting from. I assume this link is being generated in `tags/tag1/index.html` and following that leads to `tags/tag1/tags/tag1/_p/2`?

yes! So it's the user's responsibility to add the leading slash again then?

File: src/pagination/mod.rs [hunk 4]
```diff
@@ -136,23 +135,45 @@ fn interpret_permalink(
...
-    Ok(if page_num == 1 {
-        if let Some(index) = index {
-            let include: &str = config.include.into();
-            format!("{}/{}", include, index_to_string(&index))
-        } else {
-            doc.url_path.clone()
-        }
+    let mut attributes = document::permalink_attributes(&doc.front, &doc.file_path);
+    let permalink = permalink::explode_permalink(&config.front_permalink, &attributes)?;
+    let permalink_path = std::path::Path::new(&permalink);
```
Comment (epage) [inline src/pagination/mod.rs]: Not sure how I feel about using `std::path` for manipulating a URI path

File: src/pagination/mod.rs [hunk 5]
```diff
@@ -136,23 +135,45 @@ fn interpret_permalink(
...
+        || permalink.clone(),
+        |os_str| {
+            permalink
+                .trim_end_matches(&format!(".{}", os_str.to_string_lossy()))
+                .to_string()
+        },
+    );
+    // avoid "//"
+    if pagination_root.len() == 1 && pagination_root.starts_with('/') {
+        pagination_root.remove(0);
```
Comment (epage) [inline src/pagination/mod.rs]: If `explode` strips the leading `/`, why is this needed?

**Comment by epage** [comment]
> yes! So it's the user's responsibility to add the leading slash again then?

Correct.  We should fix this in the docs.

File: src/cobalt_model/permalink.rs (hunk 1 above)
Comment (Geobert) [reply src/cobalt_model/permalink.rs]: I'll revert this, I think it's the same issue of the leading slash to be added by the user

File: src/pagination/mod.rs (hunk 4 above)
Comment (Geobert) [reply src/pagination/mod.rs]: Is there another tool in the std lib to test for extension?

File: src/pagination/mod.rs (hunk 5 above)
Comment (Geobert) [reply src/pagination/mod.rs]: As said above, not needed anymore :D

File: src/pagination/mod.rs (hunk 4 above)
Comment (epage) [reply src/pagination/mod.rs]: I'd either handlroll a solution (like we've done with other url stuff) or use [`Url::path_segments`](https://docs.rs/url/1.7.2/url/struct.Url.html#method.path_segments) to at least get the last segment and then check for the extension within that.

**Comment by epage** [comment]
Looks good!  Feel free to cleanup the commits and let me know when you've pushed them

**Comment by Geobert** [comment]
@epage Done!
For the `std::path` issue, I'll think about on how to do it properly
